spring-messaging/src/jmh/java/org/springframework/messaging/simp/broker/DefaultSubscriptionRegistryBenchmark.java

```diff
@@ -0,0 +1,206 @@
+					.toArray(String[]::new);
+
+			this.sessionIds = IntStream.range(0, this.numberOfSessions)
+					.mapToObj(i -> "sessionId_" + i)
+					.toArray(String[]::new);
+
+			for (String sessionId : this.sessionIds) {
+				registerSubscriptions(sessionId, this.subscriptionsPerSession);
+			}
+		}
+
+		public void registerSubscriptions(String sessionId, int subscriptionCount) {
+			for (int subscription = 0; subscription < subscriptionCount; subscription++) {
+				String destination = getDestinationForSubscription(subscription);
+				this.registry.registerSubscription(subscribeMessage(sessionId, "subscription" + subscription, destination));
+			}
+		}
+
+		public String getDestinationForSubscription(int subscription) {
+			return this.destinations[subscription % this.destinations.length];
+		}
+	}
+
+	@State(Scope.Thread)
+	public static class Requests {
+		public static final AtomicInteger counter = new AtomicInteger();
+
+		@Param({"false", "true"})
+		boolean sameDestination;
+
+		@Param({"false", "true"})
+		boolean sameSession;
+
+		public String session;
+
+		public Message<?> subscribe;
+
+		public String findDestination;
+
+		public Message<?> unsubscribe;
+
+		@Setup(Level.Trial)
+		public void doSetup(ServerState serverState) {
+			int someInteger = counter.getAndIncrement();
+			if (this.sameSession) {
+				this.session = serverState.sessionIds[0];
+			}
+			else {
+				this.session = serverState.sessionIds[someInteger % serverState.sessionIds.length];
+			}
+			String subscription = String.valueOf(someInteger);
+			String destination;
+			if (this.sameDestination) {
+				destination = serverState.destinations[0];
+				this.findDestination = serverState.destinations[0];
+			}
+			else {
+				destination = serverState.destinations[someInteger % serverState.destinations.length];
+				this.findDestination = serverState.destinations[someInteger % serverState.destinations.length];
+			}
+			this.subscribe = subscribeMessage(this.session, subscription, destination);
+			this.unsubscribe = unsubscribeMessage(this.session, subscription);
+		}
+
+	}
+
+	@State(Scope.Thread)
+	public static class UnregisterAllSubscriptionRequest {
+
+		@Param("20")
+		public int subscriptionCount;
+
+		public String session;
+
+		@Setup(Level.Invocation)
+		public void doSetup(ServerState serverState) {
+			this.session = UUID.randomUUID().toString();
+			serverState.registerSubscriptions(this.session, this.subscriptionCount);
+		}
+	}
+
+	@State(Scope.Thread)
+	public static class FindRequest {
+		public static final AtomicInteger counter = new AtomicInteger();
+
+		public String destination;
+
+		public Message<?> message;
+
+		@Setup(Level.Trial)
+		public void doSetup(ServerState serverState) {
+			this.destination = serverState.destinations[this.counter.getAndIncrement() % serverState.destinations.length];
+			this.message = null;
+		}
+	}
+
+	@Benchmark
+	public void registerSubscription(ServerState serverState, Requests request) {
```

Comment:
Benchmark methods should either return a value or use `Blackhole` to consume values produced by the benchmark method. Without that, we're at risk that the JVM performs dead code elimination. See https://github.com/spring-projects/spring-framework/wiki/Micro-Benchmarks for more information.

The register/unregister subscriptions return `void`, but using `Blackhole` on `findSubscriptionsInternal` should do the trick.

Reply:
thanks for this finding, I will add the blackhole